orbit/pkg/useraction/mdm_migration_darwin.go

```diff
@@ -0,0 +1,289 @@
+//go:build darwin
+
+package useraction
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+type swiftDialogExitCode int
+
+const (
+	primaryBtnExitCode   = 0
+	errorExitCode        = 1
+	secondaryBtnExitCode = 2
+	infoBtnExitCode      = 3
+	timeoutExitCode      = 4
+	userQuitExitCode     = 10
+	unknownExitCode      = 99
+)
+
+var mdmMigrationTemplate = template.Must(template.New("mdmMigrationTemplate").Parse(`
+## Migrate to Fleet
+
+To begin, click "Start." Your default browser will open your My Device page.
+
+{{ if .Aggresive }}You {{ else }} Once you start, you {{ end -}} will see this dialog every 15 minutes until you click "Turn on MDM" and complete the instructions.
+
+\\![Image showing the Fleet UI](http://localhost:64793/mdm-migration-screenshot-768x180@2x.jpg)
+
+Unsure? Contact {{ .OrgInfo.OrgName }} IT [here]({{ .OrgInfo.ContactURL }}).
+`))
+
+var errorTemplate = template.Must(template.New("").Parse(`
+### Something's gone wrong.
+
+Please contact your IT admin [here]({{ .ContactURL }}).
+`))
+
+type baseDialog struct {
+	path        string
+	interruptCh chan struct{}
+}
+
+func newBaseDialog(path string) *baseDialog {
+	return &baseDialog{path: path, interruptCh: make(chan struct{})}
+}
+
+func (b *baseDialog) CanRun() bool {
+	if _, err := os.Stat(b.path); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false
+		}
+		return false
+	}
+
+	return true
+}
+
+func (b *baseDialog) Exit() {
+	b.interruptCh <- struct{}{}
+	log.Info().Msg("dialog exit message sent")
+}
+
+func (b *baseDialog) render(flags ...string) (chan swiftDialogExitCode, chan error) {
```

Comment:
It is quite complex concurrent code with various ways it is called, so it's impossible to be completely sure that it's ok not being able to run it. If you have the chance to do a run with `-race` enabled on macOS and a mocked/test swift UI, it would be ideal. I still have found a couple of issues that I'll comment on below.

Reply:
it's complex and last-minute, sorry about that! and thank you for all the insight!

I'm currently working on the tests, but decided to put the PR up given the urgency.

Reply:
No worries, totally understand. I think the fast-path to make the release, from what I can see, would be to keep it mostly as-is but use buffered channels of 1 in `render`, and ensure that at most one value is ever sent to either channel. That would take care of any blocking/goroutine leak even if the caller does not care about the return values.

Reply:
thanks! I think this also answers the question I just posted in the sibling thread